### PR TITLE
fix(mysql): DATE_ADD for datetimes

### PR DESCRIPTION
**Pull request by barakalon**
The recent logic I added for transpiling DATE_ADD in MySQL was wrong.

MySQL's DATE_ADD semantics are different than Hive's: https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-add

I'm not sure of a good way to handle this at the parser/generator level without a bunch of new TsOrDs* functions. So I added it to the optimizer for now.

### CODE REVIEW

File: tests/fixtures/optimizer/tpc-ds/tpc-ds.sql [hunk 1]
```diff
@@ -9775,7 +9775,7 @@ JOIN "date_dim" AS "d1"
   ON "catalog_sales"."cs_sold_date_sk" = "d1"."d_date_sk"
   AND "d1"."d_week_seq" = "d2"."d_week_seq"
   AND "d1"."d_year" = 2002
-  AND "d3"."d_date" > CONCAT("d1"."d_date", INTERVAL '5' day)
+  AND "d3"."d_date" > "d1"."d_date" + INTERVAL '5' day
```
Comment (barakalon) [inline tests/fixtures/optimizer/tpc-ds/tpc-ds.sql:9778]: Seems like this change caught a bug.

File: sqlglot/expressions.py [hunk 2]
```diff
@@ -234,6 +234,9 @@ def type(self) -> t.Optional[DataType]:
 
     @type.setter
     def type(self, dtype: t.Optional[DataType | DataType.Type | str]) -> None:
+        self.set_type(dtype)
+
+    def set_type(self, dtype: t.Optional[DataType | DataType.Type | str]) -> None:
```
Comment (GeorgeSittas) [inline sqlglot/expressions.py:239]: Why was this needed?

File: sqlglot/expressions.py [hunk 3]
```diff
@@ -4053,6 +4056,16 @@ def unit(self) -> t.Optional[Var]:
...
 
 
+class IntervalOp(TimeUnit):
+    arg_types = {"unit": True, "expression": True}
+
+    def interval(self):
+        return Interval(
+            this=self.expression.copy(),
+            unit=self.unit.copy(),
+        )
```
Comment (GeorgeSittas) [inline sqlglot/expressions.py:4066]: Should we make this a property and call it `to_interval` or something?

File: sqlglot/optimizer/annotate_types.py [hunk 4]
```diff
@@ -421,3 +484,14 @@ def _annotate_by_args(
             )
 
         return expression
+
+    def _annotate_dateadd(self, expression: exp.IntervalOp) -> exp.IntervalOp:
+        self._annotate_args(expression)
+
+        if expression.this.type.is_type(*exp.DataType.TEXT_TYPES):
```
Comment (GeorgeSittas) [inline sqlglot/optimizer/annotate_types.py:491]: Should we instead phrase this as `expression.this.type in exp.DataType.TEXT_TYPES` to avoid unpacking?

File: tests/test_optimizer.py [hunk 5]
```diff
@@ -546,6 +546,52 @@ def test_binary_annotation(self):
...
+                "SELECT CAST('2023-01-01' AS DATETIME) + INTERVAL '1' HOUR",
+            ),
+            (
+                "SELECT '2023-01-01 00:00:01' + INTERVAL '1' HOUR",
+                exp.DataType.Type.DATETIME,
+                "SELECT CAST('2023-01-01 00:00:01' AS DATETIME) + INTERVAL '1' HOUR",
+            ),
+            ("SELECT 'nonsense' + INTERVAL '1' DAY", exp.DataType.Type.UNKNOWN),
+            ("SELECT x.a + INTERVAL '1' DAY FROM x AS x", exp.DataType.Type.DATE),
+            ("SELECT x.a + INTERVAL '1' HOUR FROM x AS x", exp.DataType.Type.DATE),
```
Comment (GeorgeSittas) [inline tests/test_optimizer.py:574]: Why is this a date? This doesn't seem to match MySQL's behavior, for example:

```
mysql> select cast('2020-01-01' as date) + interval '1' hour;
+------------------------------------------------+
| cast('2020-01-01' as date) + interval '1' hour |
+------------------------------------------------+
| 2020-01-01 01:00:00                            |
+------------------------------------------------+
1 row in set (0.00 sec)
```

**Review by GeorgeSittas — COMMENTED** [review]
Nice work @barakalon, these date/time optimizations look really interesting. I wonder if they are valid in all supported dialects though? Date/time stuff can vary wildly...

Left a few comments to consider.

File: sqlglot/expressions.py (hunk 2 above)
Comment (barakalon) [reply sqlglot/expressions.py:239]: mypy doesn't like
```python
expression.set = DataType.Type.Foo
```
Seems to be using the type of the getter, not the setter?

File: sqlglot/expressions.py (hunk 3 above)
Comment (barakalon) [reply sqlglot/expressions.py:4066]: I'll call it to_interval

File: sqlglot/optimizer/annotate_types.py (hunk 4 above)
Comment (barakalon) [reply sqlglot/optimizer/annotate_types.py:491]: ah sure

File: tests/test_optimizer.py (hunk 5 above)
Comment (barakalon) [reply tests/test_optimizer.py:574]: Ah, you're right. This should be a datetime. good catch

**Comment by barakalon** [comment]
Yeah, I also am unsure of all the different nuances of different dialects.
I guess it should be easy to make configurable though, given the implementation with `BINARY_COERCIONS`?

File: sqlglot/expressions.py (hunk 2 above)
Comment (tobymao) [reply sqlglot/expressions.py:239]: just do type ignore then

File: sqlglot/expressions.py (hunk 3 above)
Comment (barakalon) [reply sqlglot/expressions.py:4066]: Hm, `to_interval` isn't quite right, because it doesn't operate on the entire expression. Just the interval part.
And making it a property doesn't feel right either, since this creates a new instance.

Comment (GeorgeSittas) [reply sqlglot/expressions.py:4066]: I see what you mean, feel free to ignore this then.

**Review by GeorgeSittas — APPROVED** [review]
